### PR TITLE
frontend: use navigate -1 for back buttons

### PR DESCRIPTION
**Pull request by thisconnect**
Back buttons should not add new history entries and behave same
as native back buttons by going one step back in history.

also: added functional back button component for send view class component

### CODE REVIEW

File: frontends/web/src/routes/account/receive/receive.tsx [hunk 1]
```diff
@@ -280,11 +280,11 @@ export const Receive = ({
                       primary>
                       {t('receive.verifyBitBox02')}
                     </Button>
-                    <ButtonLink
+                    <Button
```
Comment (benma) [inline frontends/web/src/routes/account/receive/receive.tsx:283]: Why don't you use the Backbutton component here and in other places?

File: frontends/web/src/routes/account/send/components/back.tsx [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { useNavigate } from 'react-router-dom';
+import { Button, TButton } from '@/components/forms/button';
+
+type TBackButton = Omit<TButton, 'onClick'>;
```
Comment (benma) [inline frontends/web/src/routes/account/send/components/back.tsx:20]: you don't actually propagate all the other props to <Button>, so why this type? you only need `children`.

File: frontends/web/src/routes/account/send/send.tsx [hunk 3]
```diff
@@ -637,11 +638,9 @@ class Send extends Component<Props, State> {
                         disabled={!this.getValidTxInputData() || !valid || isUpdatingProposal}>
                         {t('send.button')}
                       </Button>
-                      <ButtonLink
-                        secondary
-                        to={`/account/${code}`}>
+                      <BackButton secondary>
```
Comment (benma) [inline frontends/web/src/routes/account/send/send.tsx:641]: I am confused, `secondary` is hardcoded in BackButton and this `secondary` here is unused I think. Why is nothing complaining about unused props?

**Review by benma — CHANGES_REQUESTED** [review]

File: frontends/web/src/routes/account/receive/receive.tsx (hunk 1 above)
Comment (thisconnect) [reply frontends/web/src/routes/account/receive/receive.tsx:283]: BackButton is just a simple wrapper so we can use the hook in class based component. 

In one place it shares the onClick function with useEsc and I guess we do that more in the future. So should the new BackButton do this as well? what if there is another popup that listens to esc etc?

I prefer to keep it simple for now, but no strong opinion.

File: frontends/web/src/routes/account/send/components/back.tsx (hunk 2 above)
Comment (thisconnect) [reply frontends/web/src/routes/account/send/components/back.tsx:20]: good catch it should pass props.. but lets just find out if we want BackButton everywhere or if this is just a temporary component that can be removed once Send is a functional component.

File: frontends/web/src/routes/account/send/send.tsx (hunk 3 above)
Comment (thisconnect) [reply frontends/web/src/routes/account/send/send.tsx:641]: TButton accepts this prop. This is a mistake, I'll update once we decide if BackButton has a future or is just temp.

Comment (benma) [reply frontends/web/src/routes/account/send/send.tsx:641]: Imo it has a future just so the `navigate(-1)` onclick thing does not have to be repeated everywhere. then we could also easily hide it completely on Android if we wanted to.

Comment (thisconnect) [reply frontends/web/src/routes/account/send/send.tsx:641]: ack 👍

File: frontends/web/src/routes/account/send/services.ts [hunk 4]
```diff
@@ -37,8 +37,7 @@ export const txProposalErrorHandling = (registerEvents: () => void, unregisterEv
     return { feeError: t(`send.error.${errorCode}`) };
   default:
     if (errorCode) {
-      unregisterEvents();
-      alertUser(errorCode, { callback: registerEvents });
+      alertUser(errorCode);
```
Comment (thisconnect) [inline frontends/web/src/routes/account/send/services.ts:40]: this has the side-effect that if an unknown error occurred and Alert is shown the user can hit the esc key and go back.

**Comment by thisconnect** [comment]
@benma I pushed one more commit that updates BackButton and added canGoBack prop, so each component can pass a boolean if the back button should be active or disabled. Should also address the other comments about passong props.

I'm not sure if you had something like this in mind.

I keep it as separate commit and if ok I'll update all other views to use this BackButton.

Or let me know how you'd like the BackButton to work.

File: frontends/web/src/components/backbutton/backbutton.tsx [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { ComponentPropsWithoutRef, useCallback } from 'react';
+import { useNavigate } from 'react-router-dom';
+import { useEsc } from '@/hooks/keyboard';
+import { Button } from '@/components/forms/button';
+
+type TBackButton = Omit<ComponentPropsWithoutRef<'button'>, 'onClick'> & {
```
Comment (benma) [inline frontends/web/src/components/backbutton/backbutton.tsx:22]: shouldn't this be using button.tsx:TProps as you are passing props to the Button component?

File: frontends/web/src/components/backbutton/backbutton.tsx [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+  }, [canGoBack, navigate]);
+
+  useEsc(handleBack);
+
+  return (
+    <Button
+      secondary
+      onClick={handleBack}
+      disabled={!canGoBack}
+      {...props}>
```
Comment (benma) [inline frontends/web/src/components/backbutton/backbutton.tsx:45]: Looks like `disabled` can be passed in these props too or not? Could be confusing having both canGoBack and disabled. Imho I'd just remove `canGoBack` and use `disabled` instead where needed.

**Review by benma — COMMENTED** [review]
I still think having ESC behave like a back button everywhere is not normal and not intuitive. ESC in the send screen is probably a bit weird too, but one could consider send/receive as kind of dialogs on the account (even if they are not). But going e.g setings->device-settings->manage accounts, ESC going back to device settings is not intuitive at all.

Maybe just add a `enableEsc` prop to enable it in the send screen to not change current behavior for now?

File: frontends/web/src/components/backbutton/backbutton.tsx (hunk 5 above)
Comment (thisconnect) [reply frontends/web/src/components/backbutton/backbutton.tsx:22]: Yes it probably should, but I tried and could not get it to work, as TButton has a bit a special type for `primary` OR `secondary` OR `transparent`.

And BackButton should always be `secondary` anyway, and never have onClick. So I figured it is not worth bothering extending the type of TButton.

File: frontends/web/src/components/backbutton/backbutton.tsx (hunk 6 above)
Comment (thisconnect) [reply frontends/web/src/components/backbutton/backbutton.tsx:45]: makes sense

File: frontends/web/src/components/backbutton/backbutton.tsx (hunk 5 above)
Comment (benma) [reply frontends/web/src/components/backbutton/backbutton.tsx:22]: why even pass on props in the first place? maybe better just make an explicit list of props that are needed for clarity. Or maybe Pick<> instead of Omit<> for the ones that we actually use and want to pass on.

Comment (thisconnect) [reply frontends/web/src/components/backbutton/backbutton.tsx]: Pushed a commit to simplify BackButton, I think disabled, enableEsc, className and children are sufficient.

Children is probably not really necessary, but I like the flexibility so that the parent component can decide to use button.back or "cancel" abort etc. depending on the view/popup

File: frontends/web/src/components/forms/button.tsx [hunk 7]
```diff
@@ -84,13 +88,14 @@ export const Button = ({
...
-            || (danger && 'danger')
-            || 'button'
+    style[
+      (primary && 'primary')
+      || (secondary && 'secondary')
+      || (transparent && 'transparent')
+      || (danger && 'danger')
+      || 'button'
     ], className
   ].join(' ');
```
Comment (thisconnect) [inline frontends/web/src/components/forms/button.tsx:101]: I kept this changes, but should probably move to a separate commit.

File: frontends/web/src/routes/account/send/send.tsx [hunk 8]
```diff
@@ -637,11 +625,11 @@ class Send extends Component<Props, State> {
                         disabled={!this.getValidTxInputData() || !valid || isUpdatingProposal}>
                         {t('send.button')}
                       </Button>
-                      <ButtonLink
-                        secondary
-                        to={`/account/${code}`}>
+                      <BackButton
+                        disabled={activeCoinControl || activeScanQR}
+                        enableEsc>
```
Comment (thisconnect) [inline frontends/web/src/routes/account/send/send.tsx:630]: Setting the button to disabled it not really needed as the UI elements are covered by either coin-control or qr-scan popups.

Similar situation as in receive, but there it currently only conditionally sets enableEsc `<BackButton enableEsc={!addressTypeDialog && !verifying}>`

**Comment by thisconnect** [comment]
I'll make nicer commits when we are happy with the props. @benma PTAL

File: frontends/web/src/components/backbutton/backbutton.tsx [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+}: TBackButton) => {
+  const navigate = useNavigate();
+
+  const handleBack = useCallback(() => {
+    if (!disabled) {
+      navigate(-1);
+    }
+  }, [disabled, navigate]);
+
+  useEsc(() => enableEsc && handleBack());
```
Comment (benma) [inline frontends/web/src/components/backbutton/backbutton.tsx:43]: i am never quite sure how it works, but maybe you need useCallback(() => ...)) here to avoid re-renders?

**Review by benma — APPROVED** [review]
utACK

File: frontends/web/src/components/backbutton/backbutton.tsx (hunk 9 above)
Comment (thisconnect) [reply frontends/web/src/components/backbutton/backbutton.tsx]: I tested a bit and found a perf issue, but the other way around. 

If there are many re-renders this causes the internal useKeyboard hook to remove and re-attach the keyboad event listener on document.

On transaction overview this re-attached the event over 100 times (1 per tx). Added a commit to address this and not render Dialog when there is no data.
